cdap-app-fabric/src/test/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeServiceTest.java

```diff
@@ -99,10 +115,12 @@ public void testConcurrentStartLimit() throws Exception {
     ProgramDescriptor descriptor = new ProgramDescriptor(program.getId(), null,
                                                          NamespaceId.DEFAULT.artifact("test", "1.0"));
 
-    CConfiguration cConf = CConfiguration.create();
     cConf.setInt(Constants.AppFabric.PROGRAM_LAUNCH_THREADS, 2);
-    ProgramRuntimeService runtimeService = new TestProgramRuntimeService(cConf,
-                                                                         runnerFactory, program, null, null);
+    ProgramRuntimeService runtimeService = new TestProgramRuntimeService(cConf, runnerFactory,
+        null, programRunDispatcherFactory);
+    TestProgramRunDispatcher launchDispatcher = new TestProgramRunDispatcher(cConf, runnerFactory,
+        null, null, null, program);
+    when(programRunDispatcherFactory.getProgramRunDispatcher()).thenReturn(launchDispatcher);
```

Comment:
this is normally done through guice bindings in our unit tests

Reply:
Since we want to provide our implementation of Dispatcher which is defined in this test case, this is an easier way to return that dispatcher as in case of solving this with guice bindings, we'll have to also alter the factory to include one more binding for the test. Let me know if we still want to do it via guice bindings.

Reply:
Why do we need the factory layer though? It seems simpler just to bind the ProgramRunDispatcher to the right implementation, remote, local, or the custom one for tests here.

Reply:
Shouldn't be altering the real factory, but binding the factory to one that always returns the Dispatcher defined in this test case

Reply:
FYI: We're bringing back the factory logic since we need to decide Dispatcher based on ProgramType.